testsys: Support environment kubecontext

ecpullen opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #2426 

**Description of changes:**

Enables testsys to use an environments kubeconfig if the kubeconfig path doesn't exist.

**Testing done:**

Tested with kubeconfig in `testsys.kubeconfig` and with `kind create cluster`.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 commented on tools/testsys/src/main.rs:
```diff
@@ -34,16 +34,23 @@ struct TestsysArgs {
-        let client = match self.kubeconfig {
+        let kubeconfig_path = self
+            .kubeconfig
+            .filter(|kubeconfig| !self.default || kubeconfig.exists());
```
Following our offline discussion:

I didn't read the code properly. Now it makes senses but I have a different complain, even though `-d` is always set, it will be ignored if `TESTSYS_KUBECONFIG_PATH` exists, that makes me feel like the CLI isn't respecting the instructions set with the `testsys -d` command, which clearly says use the default.

If your intention is to always use `TESTSYS_KUBECONFIG_PATH` regardless of the flags, then I think what I suggested fits better: don't add a flag, and just express the precedence.

```
// pseudo-code
if path.is_some() && path.exists() {
    TestManager::from_file(&path)
} else {
     TestManager::new();
}
```

With something similar to ^, you respect the instruction given in the command:

```bash
# Use env variable
export TESTSYS_KUBECONFIG_PATH=here; testsys logs --test

# Use default
testsys logs --test
```

It is easier to document as well, otherwise the documentation will be like "if ENV is set, it will be used even if `-d` is passed" (I don't like when CLIs aren't clear with their flags :sweat_smile:)

ecpullen replied:
The only problem with that is it may not do what the user expects.
Assume my kubeconfig is `my-kubeconfig.yaml`.
If I ran `testsys --kubeconfig my_kubeconfig.yaml status` (instead of  `testsys --kubeconfig my-kubeconfig.yaml status`), it would use my default context which I clearly didn't want since I specified a kubeconfig.

arnaldo2792 replied:
Not quite, `TESTSYS_KUBECONFIG_PATH` is mapped to the `kubeconfig` option, so `self.kubeconfig` will be "something" and technically these two will be interchangeable:

```bash
# This won't use the default
testsys --kubeconfig <path-to-config> <blah>
# Neither will this
export TESTSYS_KUBECONFIG_PATH=<path-to-config> testsys <blah>
```

And only this will use `default`, without any flags passed:
```bash
# `kubeconfig` isn't set, so use the default kubeconfig
testsys <blah>
```

And ^ still mimics what other CLIs do: use `$HOME/.kube/config` unless `--kubeconfig` (or in your case `TESTSYS_KUBECONFIG_PATH`) is set.

ecpullen replied:
How would that work with `cargo make` having a default kubeconfig location set? Then cargo make would have to check to see if `TESTSYS_KUBECONFIG_PATH` exists. If it exists pass it into the command if not ignore it. Is that the implementation you would prefer? I prefer the command handling the file path logic in the CLI instead of Makefile scripts.

bcressey replied:
I agree with @arnaldo2792 - I wouldn't expect to have to pass a flag to use the default. I also would say it's more robust to have `testsys` get a known-good set of arguments, and to error out if told to use a kubeconfig that doesn't exist, since otherwise that could mask a misconfiguration.

Since `cargo make` is invoking `testsys` and sort of owns deciding how to invoke based on the repo state, it makes sense for it to test whether `${BUILDSYS_ROOT_DIR}/testsys.kubeconfig` exists, and if it doesn't, just not pass that argument.

I might like to generalize it a bit and populate `CARGO_MAKE_TESTSYS_ARGS`, and then just always pass that field into `testsys` invocations - similar to how `CARGO_MAKE_CARGO_ARGS` works.

(You can also add `TESTSYS_ARGS` at the same time so there's a variable that's ready to go for overrides that won't change the default args - similar to `CARGO_BUILD_ARGS`.)

ecpullen committed:
testsys: Support environment kubecontext

Add support for the default K8s context with `cargo make` invocations
for testsys.

ecpullen commented:
^ Updated to remove `-d` from testsys commands.

The kube context is now determined by in order:
User provided kubeconfig path `TESTSYS_KUBECONFIG_PATH`
Default kubeconfig path `testsys.kubeconfig` (if it exists it is used)
User kubecontext

Tested with a kind cluster context, eks context, setting the path with env variables and using the default path.

bcressey commented on Makefile.toml:
```diff
@@ -127,13 +127,30 @@ AMI_DATA_FILE_SUFFIX = "amis.json"
 # The commit for the last release of bottlerocket.
 TESTSYS_STARTING_COMMIT = { script = ["git describe --tag ${TESTSYS_STARTING_VERSION} --always --exclude '*' || echo 00000000"] }
 
+# Determines the kubeconfig that should be used by testsys. If no kubeconfig was provided and the
+# default kubeconfig location does not exist, use the users default kubeconfig.
+CARGO_MAKE_TESTSYS_KUBECONFIG_ARG = {script = [
+   '''
+      if [[ -n ${TESTSYS_KUBECONFIG_PATH} ]]; then
+         # If the user provides a kubeconfig path it should be used.
+         echo "--kubeconfig ${TESTSYS_KUBECONFIG_PATH}"
+      elif [[ -f "${CARGO_MAKE_DEFAULT_TESTSYS_KUBECONFIG_PATH}" ]]; then
+         # If the default kubeconfig exists it should be used.
+         echo "--kubeconfig ${CARGO_MAKE_DEFAULT_TESTSYS_KUBECONFIG_PATH}"
+      fi
+   '''
+]}
```
Is there a need for both `TESTSYS_KUBECONFIG_PATH` and `CARGO_MAKE_DEFAULT_TESTSYS_KUBECONFIG_PATH` ?

Couldn't we just keep the old default value and then they could override the kubeconfig by setting that one variable via `-e TESTSYS_KUBECONFIG_PATH`?

If you tried that and it didn't work, it might help to move this into the `env.development` block. I believe dependencies between env variables was a feature added in a very recent `cargo-make` release, but I don't quite want to rely on that yet.

ecpullen replied:
If we only use `TESTSYS_KUBECONFIG_PATH` there is no way to determine if the user was trying to override the path or if it was the default path. I'll move it to `env.development`.

stmcginnis replied:
Do we need to know that? It seems like it would be nice if this worked like some of our other overrides (e.g. `-e BUILDSYS_VARIANT`) where if not specified it just uses the default, otherwise the user can override the variable with their custom location.

ecpullen replied:
The main issue is, if the user supplies a kubeconfig, but there is a typo and it doesn't exist, it would default to using an unplanned kubeconfig which should not happen.

ecpullen commented:
^ Move `CARGO_MAKE_TESTSYS_KUBECONFIG_ARG` to `env.development`

bcressey commented on Makefile.toml:
```diff
@@ -184,6 +184,23 @@ BOOT_CONFIG_INPUT = "${BUILDSYS_ROOT_DIR}/bootconfig-input"
+# Determines the kubeconfig that should be used by testsys. If no kubeconfig was provided and the
+# default kubeconfig location does not exist, use the users default kubeconfig.
+CARGO_MAKE_TESTSYS_KUBECONFIG_ARG = {script = [
+   '''
```
nit: align these quotes and the following script to the left margin

bcressey commented on Makefile.toml:
```diff
@@ -184,6 +184,23 @@ BOOT_CONFIG_INPUT = "${BUILDSYS_ROOT_DIR}/bootconfig-input"
+# default kubeconfig location does not exist, use the users default kubeconfig.
+CARGO_MAKE_TESTSYS_KUBECONFIG_ARG = {script = [
+   '''
+      if [[ -n ${TESTSYS_KUBECONFIG_PATH} ]]; then
```
style nits: quote all shell variables and avoid `[[` if `[` will do:
```suggestion
      if [ -n "${TESTSYS_KUBECONFIG_PATH}" ]; then
```

bcressey commented on Makefile.toml:
```diff
@@ -184,6 +184,23 @@ BOOT_CONFIG_INPUT = "${BUILDSYS_ROOT_DIR}/bootconfig-input"
+      if [[ -n ${TESTSYS_KUBECONFIG_PATH} ]]; then
+         # If the user provides a kubeconfig path it should be used.
+         echo "--kubeconfig ${TESTSYS_KUBECONFIG_PATH}"
+      elif [[ -f "${CARGO_MAKE_DEFAULT_TESTSYS_KUBECONFIG_PATH}" ]]; then
```
```suggestion
      elif [ -f "${CARGO_MAKE_DEFAULT_TESTSYS_KUBECONFIG_PATH}" ]; then
```

bcressey commented on Makefile.toml:
```diff
@@ -184,6 +184,23 @@ BOOT_CONFIG_INPUT = "${BUILDSYS_ROOT_DIR}/bootconfig-input"
+CARGO_MAKE_TESTSYS_KUBECONFIG_ARG = {script = [
+   '''
+      if [[ -n ${TESTSYS_KUBECONFIG_PATH} ]]; then
+         # If the user provides a kubeconfig path it should be used.
+         echo "--kubeconfig ${TESTSYS_KUBECONFIG_PATH}"
```
Since we expect it to be overridden, `TESTSYS_KUBECONFIG` might be a better name for the variable in that it's shorter and not too ambiguous.

ecpullen commented:
^ Force push to address style nits and change `TESTSYS_KUBECONFIG_PATH` to `TESTSYS_KUBECONFIG`.

webern reviewed (approved):
Are the `CARGO_MAKE_` prefixes necessary? They make the variables wordy but maybe don't add value? Maybe there is a convention already in place that you are following...

ecpullen commented:
> Are the `CARGO_MAKE_` prefixes necessary? They make the variables wordy but maybe don't add value? Maybe there is a convention already in place that you are following...

This was recommended as those env variables are determined from user input for cargo make to use. They probably shouldn't be overridden by a user either.
